quickwit/quickwit-common/src/io.rs

```diff
@@ -0,0 +1,391 @@
+// Copyright (C) 2022 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+// This file contains code copied from the Resource trait
+// in async-speed-limit from the TiKV project.
+// https://github.com/tikv/async-speed-limit/blob/master/src/io.rs
+//
+// Copyright 2019 TiKV Project Authors. Licensed under MIT or Apache-2.0.
+
+// We are simply porting the logic to tokio here and adding the functionality to
+// plug some metrics.
+
+use std::future::Future;
+use std::io;
+use std::io::IoSlice;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+use async_speed_limit::clock::StandardClock;
+use async_speed_limit::limiter::Consume;
+use async_speed_limit::Limiter;
+use once_cell::sync::Lazy;
+use pin_project_lite::pin_project;
+use prometheus::{IntCounter, IntCounterVec};
+use tokio::io::AsyncWrite;
+
+use crate::metrics::new_counter_vec;
+use crate::{KillSwitch, Progress, ProtectedZoneGuard};
+
+// Max 1MB at a time.
+const MAX_NUM_BYTES_WRITTEN_AT_ONCE: usize = 1 << 20;
+
+fn truncate_bytes(bytes: &[u8]) -> &[u8] {
+    let num_bytes = bytes.len().min(MAX_NUM_BYTES_WRITTEN_AT_ONCE);
+    &bytes[..num_bytes]
+}
+
+struct IoMetrics {
+    write_num_bytes_total: IntCounterVec,
```

Comment:
`num_bytes_written_total`? (`num_bytes_written` is 5x more popular on GitHub codebases)

Reply:
it was to stick to prometheus convention, but actually we have a unit here.

Reply:
rename `write_bytes`. https://prometheus.io/docs/practices/naming/